File: leetcode/algorithms/66_plus_one/main.java
import java.util.Arrays;

class PlusOne {
  /**
   * Complexities:
   *   N - Length of `digits`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(N)
   */
  public int[] plusOne(int[] digits) {
    for (int i = digits.length - 1; i >= 0; i--) {
      digits[i]++;

      if (digits[i] < 10) {
        return digits;
      }

      digits[i] -= 10;
    }

    int[] result = new int[digits.length + 1];
    Arrays.fill(result, 0);
    result[0] = 1;

    return result;
  }


  // Solution
  /**
   * Complexities:
   *   N - Length of `digits`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(N)
   */
  public int[] solution(int[] digits) {
    int n = digits.length;

    for (int i = n - 1; i >= 0; i--) {
      if (digits[i] < 9) {
        digits[i]++;
        return digits;
      }

      digits[i] = 0;
    }

    int[] newNumber = new int[n + 1];
    newNumber[0] = 1;

    return newNumber;
  }
}

File: leetcode/algorithms/66_plus_one/main.cpp
#include <vector>
using namespace std;

class PlusOne {
public:
    /**
     * Complexities:
     *   N - Size of `digits`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(N)
     */
    vector<int> plusOne(vector<int>& digits) {
        vector<int> result;

        int up = 1;
        for (int i = digits.size() - 1; i >= 0; i--) {
            digits[i] += up;
            up = 0;

            if (digits[i] > 9) {
                digits[i] -= 10;
                up = 1;
            }
            result.insert(result.begin(), digits[i]);
        }

        if (up > 0) {
            result.insert(result.begin(), 1);
        }

        return result;
    }


    // Solution
    /**
     * Complexities:
     *   N - Size of `digits`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(1)
     */
    vector<int> solution(vector<int>& digits) {
        int n = digits.size();

        for (int i = n - 1; i >= 0; --i) {
            if (digits[i] < 9) {
                digits[i]++;
                return digits;
            }
            
            digits[i] = 0;
        }

        digits.insert(digits.begin(), 1);
        
        return digits;
    }
};
